File: test/test_operation.c
#include <complex.h>
#include "operation.h"
#include "aligned_memory.h"


char* test_mps_vdot()
{
	hid_t file = H5Fopen("../test/data/test_mps_vdot.hdf5", H5F_ACC_RDONLY, H5P_DEFAULT);
	if (file < 0) {
		return "'H5Fopen' in test_mps_vdot failed";
	}

	// number of lattice sites
	const int nsites = 4;
	// local physical dimension
	const long d = 3;

	// physical quantum numbers
	qnumber* qsite = aligned_alloc(MEM_DATA_ALIGN, d * sizeof(qnumber));
	if (read_hdf5_attribute(file, "qsite", H5T_NATIVE_INT, qsite) < 0) {
		return "reading physical quantum numbers from disk failed";
	}

	// virtual bond quantum numbers for 'psi'
	const long dim_bonds_psi[5] = { 1, 13, 17, 8, 1 };
	qnumber** qbonds_psi = aligned_alloc(MEM_DATA_ALIGN, (nsites + 1) * sizeof(qnumber*));
	for (int i = 0; i < nsites + 1; i++)
	{
		qbonds_psi[i] = aligned_alloc(MEM_DATA_ALIGN, dim_bonds_psi[i] * sizeof(qnumber));
		char varname[1024];
		sprintf(varname, "qbond_psi_%i", i);
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, qbonds_psi[i]) < 0) {
			return "reading virtual bond quantum numbers from disk failed";
		}
	}

	// virtual bond quantum numbers for 'chi'
	const long dim_bonds_chi[5] = { 1, 15, 20, 11, 1 };
	qnumber** qbonds_chi = aligned_alloc(MEM_DATA_ALIGN, (nsites + 1) * sizeof(qnumber*));
	for (int i = 0; i < nsites + 1; i++)
	{
		qbonds_chi[i] = aligned_alloc(MEM_DATA_ALIGN, dim_bonds_chi[i] * sizeof(qnumber));
		char varname[1024];
		sprintf(varname, "qbond_chi_%i", i);
		if (read_hdf5_attribute(file, varname, H5T_NATIVE_INT, qbonds_chi[i]) < 0) {
			return "reading virtual bond quantum numbers from disk failed";
		}
	}

	struct mps psi;
	allocate_mps(DOUBLE_COMPLEX, nsites, d, qsite, dim_bonds_psi, (const qnumber**)qbonds_psi, &psi);

	// read MPS tensors from disk
	for (int i = 0; i < nsites; i++)
	{
		// read dense tensors from disk
		struct dense_tensor a_dns;
		allocate_dense_tensor(psi.a[i].dtype, psi.a[i].ndim, psi.a[i].dim_logical, &a_dns);
		char varname[1024];
		sprintf(varname, "psi_a%i", i);
		if (read_hdf5_dataset(file, varname, H5T_NATIVE_DOUBLE, a_dns.data) < 0) {
			return "reading tensor entries from disk failed";
		}

		dense_to_block_sparse_tensor_entries(&a_dns, &psi.a[i]);

		delete_dense_tensor(&a_dns);
	}

	if (!mps_is_consistent(&psi)) {
		return "internal MPS consistency check failed";
	}

	struct mps chi;
	allocate_mps(DOUBLE_COMPLEX, nsites, d, qsite, dim_bonds_chi, (const qnumber**)qbonds_chi, &chi);

	// read MPS tensors from disk
	for (int i = 0; i < nsites; i++)
	{
		// read dense tensors from disk
		struct dense_tensor a_dns;
		allocate_dense_tensor(chi.a[i].dtype, chi.a[i].ndim, chi.a[i].dim_logical, &a_dns);
		char varname[1024];
		sprintf(varname, "chi_a%i", i);
		if (read_hdf5_dataset(file, varname, H5T_NATIVE_DOUBLE, a_dns.data) < 0) {
			return "reading tensor entries from disk failed";
		}

		dense_to_block_sparse_tensor_entries(&a_dns, &chi.a[i]);

		delete_dense_tensor(&a_dns);
	}

	if (!mps_is_consistent(&chi)) {
		return "internal MPS consistency check failed";
	}

	// compute inner product
	dcomplex s;
	mps_vdot(&chi, &psi, &s);

	dcomplex s_ref;
	if (read_hdf5_dataset(file, "s", H5T_NATIVE_DOUBLE, &s_ref) < 0) {
		return "reading dot product reference value from disk failed";
	}

	// compare
	if (cabs(s - s_ref) / cabs(s_ref) > 1e-12) {
		return "inner product does not match reference value";
	}

	delete_mps(&chi);
	delete_mps(&psi);
	for (int i = 0; i < nsites + 1; i++)
	{
		aligned_free(qbonds_chi[i]);
		aligned_free(qbonds_psi[i]);
	}
	aligned_free(qbonds_chi);
	aligned_free(qbonds_psi);
	aligned_free(qsite);

	H5Fclose(file);

	return 0;
}

File: test/test_operation.py
import numpy as np
import h5py
import pytenet as ptn
from util import interleave_complex


def mps_vdot_data():

    # random number generator
    rng = np.random.default_rng(395)

    # local physical dimension
    d = 3

    # physical quantum numbers
    qd = rng.integers(-1, 2, size=d)

    # virtual bond quantum numbers
    qD_psi = [rng.integers(-1, 2, size=Di) for Di in (1, 13, 17,  8, 1)]
    qD_chi = [rng.integers(-1, 2, size=Di) for Di in (1, 15, 20, 11, 1)]
    # ensure that leading and trailing virtual bond quantum numbers are compatible
    qD_chi[0] = qD_chi[-1] + qD_psi[0] - qD_psi[-1]

    # create random matrix product states
    psi = ptn.MPS(qd, qD_psi, fill="random", rng=rng)
    chi = ptn.MPS(qd, qD_chi, fill="random", rng=rng)
    # rescale tensors such that overall norm is of the order 1
    for i in range(psi.nsites):
        psi.A[i] *= 5
    for i in range(chi.nsites):
        chi.A[i] *= 5

    # calculate dot product <chi | psi>
    s = ptn.vdot(chi, psi)

    with h5py.File("data/test_mps_vdot.hdf5", "w") as file:
        file.attrs["qsite"] = qd
        for i, qbond in enumerate(qD_psi):
            file.attrs[f"qbond_psi_{i}"] = qbond
        for i, qbond in enumerate(qD_chi):
            file.attrs[f"qbond_chi_{i}"] = qbond
        for i, ai in enumerate(psi.A):
            # transposition due to different convention for axis ordering
            file[f"psi_a{i}"] = interleave_complex(ai.transpose((1, 0, 2)))
        for i, ai in enumerate(chi.A):
            # transposition due to different convention for axis ordering
            file[f"chi_a{i}"] = interleave_complex(ai.transpose((1, 0, 2)))
        file["s"] = interleave_complex(s)


def operator_inner_product_data():

    rng = np.random.default_rng(243)

    # physical dimension
    d = 2

    # physical quantum numbers
    qd = rng.integers(-1, 2, size=d)

    # virtual bond quantum numbers
    qD_psi = [rng.integers(-1, 2, size=Di) for Di in (1, 10, 19, 26,  7, 1)]
    qD_chi = [rng.integers(-1, 2, size=Di) for Di in (1, 17, 23, 12, 13, 1)]
    qD_op  = [rng.integers(-1, 2, size=Di) for Di in (1,  6, 15, 29, 14, 1)]
    # ensure that leading and trailing virtual bond quantum numbers are compatible
    qD_chi[0] = qD_chi[-1] + qD_psi[0] - qD_psi[-1] + qD_op[0] - qD_op[-1]

    # create random matrix product states and a random matrix product operator
    psi = ptn.MPS(qd, qD_psi, fill="random", rng=rng)
    chi = ptn.MPS(qd, qD_chi, fill="random", rng=rng)
    op  = ptn.MPO(qd, qD_op,  fill="random", rng=rng)
    # convert tensor entries to single precision
    psi.A = [a.astype(np.complex64) for a in psi.A]
    chi.A = [a.astype(np.complex64) for a in chi.A]
    op.A  = [a.astype(np.complex64) for a in  op.A]
    # rescale tensors such that overall norm is of the order 1
    for i in range(psi.nsites):
        psi.A[i] *= 5
    for i in range(chi.nsites):
        chi.A[i] *= 5
    for i in range(op.nsites):
        op.A[i] *= 5

    # calculate inner product <chi | op | psi>
    s = ptn.operator_inner_product(chi, op, psi)

    with h5py.File("data/test_operator_inner_product.hdf5", "w") as file:
        file.attrs["qsite"] = qd
        for i, qbond in enumerate(qD_psi):
            file.attrs[f"qbond_psi_{i}"] = qbond
        for i, qbond in enumerate(qD_chi):
            file.attrs[f"qbond_chi_{i}"] = qbond
        for i, qbond in enumerate(qD_op):
            file.attrs[f"qbond_op_{i}"] = qbond
        for i, a in enumerate(psi.A):
            # transposition due to different convention for axis ordering
            file[f"psi_a{i}"] = interleave_complex(a.transpose((1, 0, 2)))
        for i, a in enumerate(chi.A):
            # transposition due to different convention for axis ordering
            file[f"chi_a{i}"] = interleave_complex(a.transpose((1, 0, 2)))
        for i, a in enumerate(op.A):
            # transposition due to different convention for axis ordering
            file[f"op_a{i}"] = interleave_complex(a.transpose((2, 0, 1, 3)))
        file["s"] = interleave_complex(s)


def main():
    mps_vdot_data()
    operator_inner_product_data()


if __name__ == "__main__":
    main()
